### PR TITLE
[feature] sql-block-rule: add partition_num, tablet_num, cardinality in SqlBlockRule to block big/slow sql

### LINKED ISSUE #7202: [Sql Block] add a new rule to block sql by num of partition/tablet/cardinality will be scanned.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

According to the design of the sql block rule, the query that matches the certain PATTERN or sql_HASH will be blocked.
But there are still lots of big query can be executed.

I think it is a good idea that block the big sql by nums of partition/tablet/cardinality that will be scanned. For exmaple, we can set a rule for a user:
`CREATE SQL_BLOCK_RULE test_rule PROPERTIES("partition"="31", "tablet"="496", "cardinality"="1000000000")`
`SET PROPERTY [FOR 'test_user'] 'sql_block_rules' = 'test_rule'`
When a query's num of partition/tablet/cardinality is larger than the quota set in the block rules, it will throw an exception and return before executing.

### Use case

Big sql appears when:
1.  a new sql which is handwritten and different from the ones online.
2. already added partition key, but in some scenarios, such as using concat to process partition key, it didn't hit partition and did full table scan.

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Henry2SS**
Add partitionNum, tabletNum, cardinality in SqlBlockRule to block large/slow sql.

	1. set partitionNum, tabletNum, cardinality as limitations to block sqls
	2. compatible with lower version
	3. add unit tests
	4. add docs

## Proposed changes

There are many large queries that scan the entire table or large amounts of data, and most of them are triggered manually or accidentally. In any case, this poses a great threat to the stability of the cluster.
Therefore, we need more precise SQL block methods. 

It should :
1. Just bolck query statement 
2. Let the explain statement execute
3. Block large queries before being executed, but after being planned to get core information of OlapScanNode 

Close related #7202 (replace it with issue number if it exists).

Describe the overview of changes, and introduce why we need it.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #7202 ) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/blockrule/SqlBlockRuleMgr.java [hunk 1]
```diff
@@ -107,12 +108,22 @@ public void alterSqlBlockRule(AlterSqlBlockRuleStmt stmt) throws DdlException {
...
             }
             SqlBlockRule originRule = nameToSqlBlockRuleMap.get(ruleName);
+            SqlBlockUtil.checkAlterValidate(sqlBlockRule, originRule);
             if (StringUtils.isEmpty(sqlBlockRule.getSql())) {
                 sqlBlockRule.setSql(originRule.getSql());
             }
             if (StringUtils.isEmpty(sqlBlockRule.getSqlHash())) {
                 sqlBlockRule.setSqlHash(originRule.getSqlHash());
             }
+            if (StringUtils.isEmpty(sqlBlockRule.getPartitionNum().toString())) {
```
[the remark this answers is not in the document]
Comment (Henry2SS) [reply fe/fe-core/src/main/java/org/apache/doris/blockrule/SqlBlockRuleMgr.java:118]: NO. 
`sqlBlockRule ` in this place, is from `SqlBlockRule sqlBlockRule = SqlBlockRule.fromAlterStmt(stmt);`
and `stmt` is an instance of AlterSqlBlockRuleStmt.
In the method called `setProperties` in AlterSqlBlockRuleStmt.java, all the properties are set default values.

And I've tested ALTER:
1. from lower version to this version, it is compatible；
2. ALTER operation work well 
3. ALTER operatio can report relevant exceptions

File: docs/zh-CN/administrator-guide/block-rule/sql-block.md [hunk 2]
```diff
@@ -26,15 +26,23 @@ under the License.
...
 ## 规则
 
 对SQL规则增删改查
 - 创建SQL阻止规则
-    - sql：匹配规则(基于正则匹配,特殊字符需要转译)，可选
-    - sqlHash: sql hash值，用于完全匹配，我们会在`fe.audit.log`打印这个值，可选，这个参数和sql只能二选一,如果给空值，是空字符串不是null
+    - sql：匹配规则(基于正则匹配,特殊字符需要转译)，可选，默认值为 "NULL"
+    - sqlHash: sql hash值，用于完全匹配，我们会在`fe.audit.log`打印这个值，可选，这个参数和sql只能二选一，默认值为 "NULL"
+    - partitionNum: 一个扫描节点会扫描的最大partition数量，默认值为0L
+    - tabletNum: 一个扫描节点会扫描的最大扽tablet数量，默认值为0L
```
[the remark this answers is not in the document]
Comment (Henry2SS) [reply docs/zh-CN/administrator-guide/block-rule/sql-block.md:44]: Modified.

File: docs/zh-CN/administrator-guide/block-rule/sql-block.md [hunk 3]
```diff
@@ -51,15 +59,27 @@ PROPERTIES(
 mysql> select * from order_analysis;
 ERROR 1064 (HY000): errCode = 2, detailMessage = sql match regex sql block rule: order_analysis_rule
 ```
+
+```sql
+CREATE SQL_BLOCK_RULE test_rule2 PROPERTIES("partitionNum" = "30", "cardinality"="10000000000","global"="false","enable"="true")
```
[the remark this answers is not in the document]
Comment (Henry2SS) [reply docs/zh-CN/administrator-guide/block-rule/sql-block.md:64]: Added explanation.

[the remark this answers is not in the document]
Comment (Henry2SS) [reply docs/zh-CN/administrator-guide/block-rule/sql-block.md]: Modified.

**Review by morningman — APPROVED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 4]
```diff
@@ -357,8 +359,17 @@ public void execute(TUniqueId queryId) throws Exception {
...
                         return;
                     }
-                    // limitations: partitionNum, tabletNum, cardinality
+                    // limitations: partition_num, tablet_num, cardinality
+                    List<ScanNode> scanNodeList = planner.getScanNodes();
+                    for (ScanNode scanNode : scanNodeList) {
+                        if (scanNode instanceof OlapScanNode) {
+                            OlapScanNode olapScanNode = (OlapScanNode) scanNode;
+                            Catalog.getCurrentCatalog().getSqlBlockRuleMgr().checkLimitaions(olapScanNode.getSelectedPartitionNum().longValue(),
+                                        olapScanNode.getSelectedTabletsNum(), olapScanNode.getCardinality(), analyzer.getQualifiedUser());
```
Comment (DarvenDuanXJ) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:368]: I think analyzer.getQualifiedUser() can be replaced by context.getQualifiedUser()

Comment (stalary) [reply fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:368]: Ok, I will fix it in my PR for a Bugfix.

**Pull request merged**